### ISSUE TITLE
proximity.closest_point is not releasing memory

### ISSUE THREAD

**Issue by jchibane** [mikedh/trimesh#806]
Hi all,

I am processing a bunch of meshes with proximity.closest_point (also tried proximity.signed_distance). For every mesh, I have 100k points in a bounding box, for which I want to know the distance to the mesh.

I use multiprocessing for this and this method slowly eats up all memory until none is left.

Is there a way to get around this memory problem? Also, is there a way (using the soft dependencies?) to get faster execution? Currently, it seems a bit slow.
Also, a pointer to a different implementation of that functionality is welcome. Everything that might help... :)

Thanks! Great library!
Best,
Julian

**Comment by bugerry87** [comment]
### Hi @mikedh ,

I also confirm that closest_point is very slow.
I implemented a KDTree solution from scratch and be as double as fast, though numpy only!

### Trimesh

```
Model size: (30000, 3)
Query size: (30000, 3)

Mesh setup time: 0.00498652458190918

Query time: 99.0484869480133
Mean loss: 0.01397229295901547
```

### My KDTree
```
Model size: (30000, 3)
Query size: (30000, 3)

0%                                        |50%                                   |100% explored
################################
Query time: 40.181270122528076
Mean loss: 0.004614735096702003
```

I achieved this by avoiding `sqrt` completely and lazyness (avoid un-required calculations).
For next steps, I consider to accelerate my approach with numba, or end up with a C++ backend.
As soon my Lib is stable I will set it public and share to you.

Furthermore, I also implemented branching by multiprocessing.
Here the same task with 4 jobs:

### My KDTree 4 jobs

```
Model size: (30000, 3)
Query size: (30000, 3)

0%                                        |50%                                   |100% explored
#################################
Query time: 20.91393542289734
Mean loss: 0.004216524690457758
```

### Hi @jchibane ,

As I remember `multiprocessing` has indeed memory issues in specific cases.
I suggest to initialize the jobs with `multiprocessing.Pool`. (At least works for me)

Best,
Gerald

**Comment by bugerry87** [comment]
## Hi @sidmani 

Here you can give it a try: https://github.com/bugerry87/meshtree
I extracted it from my research project just today, so no guarantees.
Feel free to post issues or contribute.

Some issues are known, i.e. the exhaustively recursive trace should be refactored for sure, etc.

**Comment by sidmani** [comment]
@bugerry87 Thank you, I'll take a look. Do you have an example for converting from a trimesh object? Can I just pass mesh.vertices and mesh.faces?

**Comment by mikedh** [comment]
Yeah happy to improve/replace `closest_point` if there's a faster/better implementation! That line you highlighted is the `rtree` broad-phase checks which are indeed the slowest part.

**Comment by sidmani** [comment]
@mikedh I don't know if it would make sense to import CGAL since the bindings don't have a proper pypi package; you have to compile them from scratch. However, if there is a way to do it, CGAL is about 100x faster than the existing implementation.

**Comment by mikedh** [comment]
Haha 2 orders of magnitude is impressive, and CGAL is a great project! Which bindings were you using?

**Comment by sidmani** [comment]
https://github.com/CGAL/cgal-swig-bindings

One strange thing is that on my machine the bindings failed to build against the master branch of CGAL; I used the 5.0 release instead.

In order to do mesh distance calculations:
```
from CGAL.CGAL_Kernel import Point_3
from CGAL.CGAL_Kernel import Triangle_3
from CGAL.CGAL_AABB_tree import AABB_tree_Triangle_3_soup

    triangles = []
    for tri in mesh.triangles:
        a = Point_3(tri[0][0], tri[0][1], tri[0][2])
        b = Point_3(tri[1][0], tri[1][1], tri[1][2])
        c = Point_3(tri[2][0], tri[2][1], tri[2][2])
        triangles.append(Triangle_3(a, b, c))
    tree = AABB_tree_Triangle_3_soup(triangles)
    tree.accelerate_distance_queries()

    tree.squared_distance(Point_3(1.0, 2.0, 3.0))
```

**Comment by bugerry87** [comment]
_Note_:
Since AABBTree seems to be the key, I thought https://pypi.org/project/aabbtree/ could be an option instead of binding the CGAL lib which has almost the same purpose as trimesh. However, `aabbtree` is solving everything in OOP manner instead of exploiting the performance of fixed (numpy) arrays. So it is to expect that this solution will also exhaust time and memory.

**Comment by jchibane** [comment]
> For anyone finding this issue: [CGAL/cgal](https://github.com/CGAL/cgal) has python bindings and can do mesh distance queries using an AABB tree.

This sounds amazing. I'll try this out in the next months and report back. This would be the end of a long and exhaustive search - for me and my algorithm. :)

**Comment by mikedh** [comment]
It would be cool to benchmark IGL as well, they seem like a great project and possibly easier to install: https://libigl.github.io/libigl-python-bindings/igl_docs/#signed_distance

**Comment by jchibane** [comment]
Cool, sure, I'll try both - thanks! 
Actually I don't need the signed distance and hope the sign is computed by the normal of the triangle and is not additionally globally closing potentially open surfaces.

**Comment by jchibane** [comment]
Okay, I compared igl, pymesh, and trimesh, on a mesh of a car and 100.000 points in the near of the surface boundary:

In [15]: %timeit igl.signed_distance(boundary_points,mesh.vertices, mesh.faces)                                                                       
693 ms ± 14.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [24]: %timeit np.sqrt(pymesh.distance_to_mesh(py_mesh, boundary_points)[0])                                                                        
55.8 s ± 160 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [26]: %timeit trimesh.proximity.closest_point(mesh, boundary_points)                                                                               
Killed
(took eternity before it was killed)

So, igl is extremely fast and was easy to install. I also checked accuracy - very good as well.
Thanks for that pointer, I will use that library - thread can be closed! :+1: :)

I think it would be a good choice for the trimesh lib aswell.

Best,
Julian
